File: util/cpp/proxy/Protocol.cpp
/* 
 * Protocol.cpp
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2016 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 *
 * Author: Giacomo Benincasa    (gbenincasa@ihmc.us)
 * Created on May 29, 2015, 4:43 PM
 */

#include "Protocol.h"

#include "Logger.h"

using namespace NOMADSUtil;

const String Protocol::REGISTER_PROXY = "RegisterProxy";
const String Protocol::REGISTER_PROXY_CALLBACK = "RegisterProxyCallback";

SimpleCommHelper2::Error Protocol::doHandshake (SimpleCommHelper2 *pCommHelper, const String &service, const String &version)
{
    const char *pszMethodName = "Protocol::doHandshake";
    SimpleCommHelper2::Error error = SimpleCommHelper2::None;
    pCommHelper->sendLine (error, "%s %s", service.c_str() , version.c_str());
    if (error != SimpleCommHelper2::None) {
        return error;
    }

    // Check whether the client is expecting to connect to this service, and that
    // the protocol is the same
    int iCount = 0;
    const char **ppszBuf = pCommHelper->receiveParsed (error, &iCount);
    if (error != SimpleCommHelper2::None) {
        return error;
    }
    if (iCount != 2) {
        return SimpleCommHelper2::ProtocolError;
    }

    if ((service != ppszBuf[0]) || (version != ppszBuf[1])) {
        if (pLogger != NULL) pLogger->logMsg (pszMethodName, Logger::L_Warning,
                        "service mismatch: received %s %s, while expecting %s %s.\n",
                        ppszBuf[0], ppszBuf[1], service.c_str() , version.c_str());
        return SimpleCommHelper2::ProtocolError;
    }
    return error;
}

File: nms/src/main/java/us/ihmc/nms/Protocol.java
/*
 * Protocol.java
 *
 * This file is part of the IHMC Util Library
 * Copyright (c) 1993-2014 IHMC.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * version 3 (GPLv3) as published by the Free Software Foundation.
 *
 * U.S. Government agencies and organizations may redistribute
 * and/or modify this program under terms equivalent to
 * "Government Purpose Rights" as defined by DFARS 
 * 252.227-7014(a)(12) (February 2014).
 *
 * Alternative licenses that allow for use within commercial products may be
 * available. Contact Niranjan Suri at IHMC (nsuri@ihmc.us) for details.
 */

package us.ihmc.nms;

/**
 *
 * @author Giacomo Benincasa    (gbenincasa@ihmc.us)
 */
public class Protocol {
    static final String BROADCAST_METHOD = "broadcast";
    static final String TRANSMIT_METHOD = "transmit";
    static final String TRANSMIT_RELIABLE_METHOD = "transmit-reliable";
    static final String SET_TRANSMISSION_TIMEOUT_METHOD = "setRetransmissionTimeout";
    static final String SET_PRIMARY_INTERFACE_METHOD = "setPrimaryInterface";
    static final String START_METHOD = "start";
    static final String STOP_METHOD = "stop";
    static final String GET_MIN_MTU_METHOD = "getMinMTU";
    static final String GET_ACTIVE_NIC_AS_STRING_METHOD = "getActiveNICsForDstAddr";
    static final String GET_PROPAGATION_MODE_METHOD = "getPropagationMode";
    static final String GET_DELIVERY_QUEUE_SIZE_METHOD = "getDeliveryQueueSize";
    static final String GET_ENCRYPTION_KEY_HASH_METHOD = "getEncryptionKeyHash";
    static final String GET_RECEIVE_RATE_METHOD = "getReceiveRate";
    static final String GET_TRANSMISSION_QUEUE_SIZE_METHOD = "getTransmissionQueueSize";
    static final String GET_RESCALED_TRANSMISSION_QUEUE_MAX_SIZE_METHOD = "getRescTransmissionQueueSize";
    static final String GET_TRANSMISSION_QUEUE_MAX_SIZE_METHOD = "getTransmissionQueueMaxSize";
    static final String GET_TRANSMISSION_RATE_LIMIT_METHOD = "getTransmitRateLimit";
    static final String SET_TRANSMISSION_QUEUE_MAX_SIZE_METHOD = "setTransmissionQueueMaxSize";
    static final String SET_TRANSMIT_RATE_LIMIT_METHOD = "setTransmitRateLimit";
    static final String GET_LINK_CAPACITY_METHOD = "getLinkCapacity";
    static final String SET_LINK_CAPACITY_METHOD = "setLinkCapacity";
    static final String GET_NEIGHBOR_QUEUE_LENGTH_METHOD = "getNeighborQueueLength";
    static final String CLEAR_TO_SEND_METHOD = "clearToSend";
    static final String PING_METHOD = "ping";
}
